// thinkjs/lib/config/sys/transform.js
'use strict';

/**
 * get function
 * @param  {mixed} value  []
 * @param  {String} config []
 * @return {}        []
 */
exports.__esModule = true;
var getFn = function getFn(value, config) {
  if (!value || think.isFunction(value)) {
    return value;
  }
  value = global[value];
  if (think.isFunction(value)) {
    return value;
  }
  var msg = think.locale('CONFIG_NOT_FUNCTION', config);
  throw new Error(msg);
};
/**
 * config transform
 * @type {Object}
 */
exports['default'] = {
  post: {
    /**
     * json content type list
     * @param  {Array} value []
     * @return {Array}       []
     */
    json_content_type: function json_content_type(value) {
      if (think.isString(value)) {
        return [value];
      }
      if (!think.isArray(value)) {
        var msg = think.locale('CONFIG_NOT_VALID', 'post.json_content_type');
        throw new Error(msg);
      }
      return value;
    }
  },
  /**
   * sub doamin deploy
   * @param  {Object} value []
   * @return {Object}       []
   */
  subdomain: function subdomain(value) {
    if (think.isString(value)) {
      return { value: value };
    }
    if (think.isArray(value)) {
      var _ret = (function () {
        var obj = {};
        value.forEach(function (item) {
          obj[item] = item;
        });
        return {
          v: obj
        };
      })();

      if (typeof _ret === 'object') return _ret.v;
    }
    if (!think.isObject(value)) {
      var msg = think.locale('CONFIG_NOT_VALID', 'subdomain');
      throw new Error(msg);
    }
    return value;
  },
  /**
   * deby module list
   * @param  {Array} value []
   * @return {Array}       []
   */
  deny_module_list: function deny_module_list(value) {
    if (think.isString(value)) {
      return [value];
    }
    if (!think.isArray(value)) {
      var msg = think.locale('CONFIG_NOT_VALID', 'deny_module_list');
      throw new Error(msg);
    }
    return value;
  },
  /**
   * output content
   * @param  {String} value []
   * @return {Function}       []
   */
  output_content: function output_content(value) {
    return getFn(value, 'output_content');
  },
  /**
   * create server
   * @param  {String} value []
   * @return {Function}       []
   */
  create_server: function create_server(value) {
    return getFn(value, 'create_server');
  },
  /**
   * html cache config
   * @type {Object}
   */
  html_cache: {
    rules: function rules(_rules) {
      var data = {};
      for (var key in _rules) {
        var value = _rules[key];
        key = key.replace(/\:/g, '/');
        //to array
        if (!think.isArray(value)) {
          value = [value];
        }
        if (think.isFunction(value[1])) {
          value[2] = value[1];
          value[1] = 0;
        }
        data[key] = value;
      }
      return data;
    }
  },
  /**
   * cache config
   * @type {Object}
   */
  cache: {
    type: function type(value) {
      return value.toLowerCase();
    }
  },
  /**
   * session config
   * @type {Object}
   */
  session: {
    type: function type(value) {
      return value.toLowerCase();
    }
  }
};
module.exports = exports['default'];
