// thinkjs/lib/adapter/db/base.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _Object$keys = require('babel-runtime/core-js/object/keys')['default'];

var _Promise = require('babel-runtime/core-js/promise')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

exports.__esModule = true;

var _parseJs = require('./_parse.js');

var _parseJs2 = _interopRequireDefault(_parseJs);

/**
 * db base class
 * @type {Class}
 */

var _default = (function (_Parse) {
  _inherits(_default, _Parse);

  function _default() {
    _classCallCheck(this, _default);

    _Parse.apply(this, arguments);
  }

  /**
   * init
   * @return {} []
   */

  _default.prototype.init = function init(config) {
    _Parse.prototype.init.call(this, config);
    this.sql = '';
    this.lastInsertId = 0;
    this._socket = null;
    this.transTimes = 0; //transaction times
  };

  /**
   * get socket instance, override by sub class
   * @return {Object} [socket instance]
   */

  _default.prototype.socket = function socket() {};

  /**
   * insert data
   * @param  {Object} data    []
   * @param  {Object} options []
   * @param  {Boolean} replace []
   * @return {Promise}         []
   */

  _default.prototype.add = function add(data, options, replace) {
    var values = [];
    var fields = [];
    for (var key in data) {
      var val = data[key];
      val = this.parseValue(val);
      if (think.isString(val) || think.isBoolean(val) || think.isNumber(val)) {
        values.push(val);
        fields.push(this.parseKey(key));
      }
    }
    var sql = replace ? 'REPLACE' : 'INSERT';
    sql += ' INTO ' + this.parseTable(options.table) + ' (' + fields.join(',') + ')';
    sql += ' VALUES (' + values.join(',') + ')';
    sql += this.parseLock(options.lock) + this.parseComment(options.comment);
    return this.execute(sql);
  };

  /**
   * insert multi data
   * @param  {Array} data    [data list]
   * @param  {Object} options []
   * @param  {Boolean} replace []
   * @return {Promise}         []
   */

  _default.prototype.addMany = function addMany(data, options, replace) {
    var _this = this;

    var fields = _Object$keys(data[0]).map(function (item) {
      return _this.parseKey(item);
    }).join(',');
    var values = data.map(function (item) {
      var value = [];
      for (var key in item) {
        var val = item[key];
        val = _this.parseValue(val);
        if (think.isString(val) || think.isBoolean(val) || think.isNumber(val)) {
          value.push(val);
        }
      }
      return '(' + value.join(',') + ')';
    }).join(',');
    var sql = replace ? 'REPLACE' : 'INSERT';
    sql += ' INTO ' + this.parseTable(options.table) + '(' + fields + ')';
    sql += ' VALUES ' + values;
    sql += this.parseLock(options.lock) + this.parseComment(options.comment);
    return this.execute(sql);
  };

  /**
   * select data
   * @param  {String} fields  []
   * @param  {String} table   []
   * @param  {Object} options []
   * @return {Promise}         []
   */

  _default.prototype.selectAdd = function selectAdd(fields, table) {
    var _this2 = this;

    var options = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

    if (think.isString(fields)) {
      fields = fields.split(/\s*,\s*/);
    }
    fields = fields.map(function (item) {
      return _this2.parseKey(item);
    });
    var sql = 'INSERT INTO ' + this.parseTable(table) + ' (' + fields.join(',') + ') ';
    sql += this.buildSelectSql(options);
    return this.execute(sql);
  };

  /**
   * delete data
   * @param  {Object} options []
   * @return {Promise}         []
   */

  _default.prototype['delete'] = function _delete(options) {
    var sql = ['DELETE FROM ', this.parseTable(options.table), this.parseWhere(options.where), this.parseOrder(options.order), this.parseLimit(options.limit), this.parseLock(options.lock), this.parseComment(options.comment)].join('');
    return this.execute(sql);
  };

  /**
   * update data
   * @param  {Object} data    []
   * @param  {Object} options []
   * @return {Promise}         []
   */

  _default.prototype.update = function update(data, options) {
    var sql = ['UPDATE ', this.parseTable(options.table), this.parseSet(data), this.parseWhere(options.where), this.parseOrder(options.order), this.parseLimit(options.limit), this.parseLock(options.lock), this.parseComment(options.comment)].join('');
    return this.execute(sql);
  };

  /**
   * select
   * @param  {Object} options []
   * @return {Promise}         []
   */

  _default.prototype.select = function select(options, cache) {
    var _this3 = this;

    var sql = undefined;
    if (think.isObject(options)) {
      sql = this.buildSelectSql(options);
      cache = options.cache || cache;
    } else {
      sql = options;
    }
    if (!think.isEmpty(cache) && this.config.cache.on) {
      var key = cache.key || think.md5(sql);
      return think.cache(key, function () {
        return _this3.query(sql);
      }, cache);
    }
    return this.query(sql);
  };

  /**
   * escape string
   * @param  {String} str []
   * @return {String}     []
   */

  _default.prototype.escapeString = function escapeString(str) {
    if (!str) {
      return '';
    }
    return str.replace(/[\0\n\r\b\t\\\'\"\x1a]/g, function (s) {
      switch (s) {
        case '\0':
          return '\\0';
        case '\n':
          return '\\n';
        case '\r':
          return '\\r';
        case '\b':
          return '\\b';
        case '\t':
          return '\\t';
        case '\x1a':
          return '\\Z';
        default:
          return '\\' + s;
      }
    });
  };

  /**
   * get last sql
   * @return {String}       []
   */

  _default.prototype.getLastSql = function getLastSql() {
    return this.sql;
  };

  /**
   * get last insert id
   * @return {String} []
   */

  _default.prototype.getLastInsertId = function getLastInsertId() {
    return this.lastInsertId;
  };

  /**
   * query string
   * @param  string str
   * @return promise
   */

  _default.prototype.query = function query(sql) {
    var _this4 = this;

    this.sql = sql;
    return think.await(sql, function () {
      return _this4.socket(sql).query(sql).then(function (data) {
        return _this4.bufferToString(data);
      });
    });
  };

  /**
   * buffer to string
   * @param  {Array} data []
   * @return {Array}      []
   */

  _default.prototype.bufferToString = function bufferToString(data) {
    if (!this.config.buffer_tostring || !think.isArray(data)) {
      return data;
    }
    for (var i = 0, _length = data.length; i < _length; i++) {
      for (var key in data[i]) {
        if (think.isBuffer(data[i][key])) {
          data[i][key] = data[i][key].toString();
        }
      }
    }
    return data;
  };

  /**
   * execute sql
   * @param  {String} sql []
   * @return {}     []
   */

  _default.prototype.execute = function execute(sql) {
    var _this5 = this;

    this.sql = sql;
    return this.socket(sql).execute(sql).then(function (data) {
      if (data.insertId) {
        _this5.lastInsertId = data.insertId;
      }
      return data.affectedRows || 0;
    });
  };

  /**
   * start transaction
   * @return {Promise} []
   */

  _default.prototype.startTrans = function startTrans() {
    if (this.transTimes === 0) {
      this.transTimes++;
      return this.execute('START TRANSACTION');
    }
    this.transTimes++;
    return _Promise.resolve();
  };

  /**
   * commit
   * @return {Promise} []
   */

  _default.prototype.commit = function commit() {
    if (this.transTimes > 0) {
      this.transTimes = 0;
      return this.execute('COMMIT');
    }
    return _Promise.resolve();
  };

  /**
   * rollback
   * @return {Promise} []
   */

  _default.prototype.rollback = function rollback() {
    if (this.transTimes > 0) {
      this.transTimes = 0;
      return this.execute('ROLLBACK');
    }
    return _Promise.resolve();
  };

  /**
   * close connect
   * @return {} []
   */

  _default.prototype.close = function close() {
    if (this._socket) {
      this._socket.close();
      this._socket = null;
    }
  };

  return _default;
})(_parseJs2['default']);

exports['default'] = _default;
module.exports = exports['default'];
