// thinkjs/lib/adapter/cache/memcache.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _Object$keys = require('babel-runtime/core-js/object/keys')['default'];

exports.__esModule = true;
var MemcacheSocket = think.adapter('socket', 'memcache');

/**
 * memcache cache
 */

var _default = (function (_think$adapter$base) {
  _inherits(_default, _think$adapter$base);

  function _default() {
    _classCallCheck(this, _default);

    _think$adapter$base.apply(this, arguments);
  }

  /**
   * init
   * @param  {Object} options []
   * @return {}         []
   */

  _default.prototype.init = function init(options) {
    this.options = think.extend({}, think.config('cache'), options);
    this.timeout = this.options.timeout || 0;
    this.prefix = this.options.repfix || '';
  };

  /**
   * get memcache instance
   * @param  {String} command []
   * @return {}         []
   */

  _default.prototype.getMemcacheInstance = function getMemcacheInstance(name) {
    var options = this.parseConfig(think.config('memcache'), this.options, {
      command: name,
      from: 'cache'
    });
    this.timeout = options.timeout || this.timeout;
    this.prefix = options.prefix || this.prefix;
    return MemcacheSocket.getInstance(options, thinkCache.MEMCACHE, ['command', 'from']);
  };

  /**
   * get data
   * @param  {String} name []
   * @return {Promise}      []
   */

  _default.prototype.get = function get(name) {
    var instance = this.getMemcacheInstance('get');
    return instance.get(this.prefix + name).then(function (value) {
      if (value) {
        return JSON.parse(value);
      }
    })['catch'](function () {});
  };

  /**
   * set data
   * @param {String} name    []
   * @param {Mixed} value   []
   * @param {Number} timeout []
   */

  _default.prototype.set = function set(name, value) {
    var timeout = arguments.length <= 2 || arguments[2] === undefined ? this.timeout : arguments[2];

    if (think.isObject(name)) {
      timeout = value || timeout;
      var key = _Object$keys(name)[0];
      value = name[key];
      name = key;
    }
    var instance = this.getMemcacheInstance('set');
    var data = JSON.stringify(value);
    return instance.set(this.prefix + name, data, timeout)['catch'](function () {});
  };

  /**
   * delete data
   * @param  {String} name []
   * @return {Promise}      []
   */

  _default.prototype['delete'] = function _delete(name) {
    var instance = this.getMemcacheInstance('delete');
    return instance['delete'](this.prefix + name)['catch'](function () {});
  };

  return _default;
})(think.adapter.base);

exports['default'] = _default;
module.exports = exports['default'];
