// thinkjs/lib/util/auto_reload.js
'use strict';

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

exports.__esModule = true;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

//can not use import from
var sys_module = require('module');

var NODE_MODULES = _path2['default'].sep + 'node_modules' + _path2['default'].sep;

/**
 * auto reload file
 */

var _default = (function (_think$base) {
  _inherits(_default, _think$base);

  function _default() {
    _classCallCheck(this, _default);

    _think$base.apply(this, arguments);
  }

  /**
   * init
   * @param  {String}   srcPath  [source path]
   * @param  {Function} callback [when file has changed, callback will be invoke]
   * @param  {Boolean}  log      [log reload file]
   * @return {}            []
   */

  _default.prototype.init = function init(srcPath, callback, showLog) {
    this.srcPath = srcPath;
    this.callback = callback;
    this.showLog = showLog;
    this.prevFilesCount = 0;
  };

  /**
   * log file
   * @param  {String} file []
   * @return {}      []
   */

  _default.prototype.log = function log(file) {
    if (!this.showLog || !file) {
      return;
    }
    //only log app files changed
    if (file.indexOf(think.srcPath) === 0) {
      file = file.slice(think.srcPath.length);
      think.log('reload file ' + file, 'RELOAD');
    }
  };

  /**
   * clear file cache, also clear dependents file cache
   * @return {} []
   */

  _default.prototype.clearFileCache = function clearFileCache(file) {
    var mod = require.cache[file];
    if (!mod) {
      return;
    }
    //remove children
    if (mod && mod.children) {
      mod.children.length = 0;
    }

    // clear module cache which dependents this module
    for (var fileItem in require.cache) {
      if (fileItem === file || fileItem.indexOf(NODE_MODULES) > -1) {
        continue;
      }
      var item = require.cache[fileItem];
      if (item && item.children && item.children.indexOf(mod) > -1) {
        this.clearFileCache(fileItem);
      }
    }
    //remove require cache
    delete require.cache[file];
  };

  /**
   * clear files cache
   * @param  {Array} files []
   * @return {}       []
   */

  _default.prototype.clearFilesCache = function clearFilesCache(files) {
    var _this = this;

    files.forEach(function (file) {
      _this.clearFileCache(file);
    });
    if (this.callback) {
      this.callback();
    }
  };

  /**
   * check file change
   * compare files count
   * @return {} []
   */

  _default.prototype.checkFileChange = function checkFileChange() {
    var filesCount = think.getFiles(this.srcPath).filter(function (file) {
      var extname = _path2['default'].extname(file);
      return extname === '.js';
    }).length;
    var flag = this.prevFilesCount && this.prevFilesCount !== filesCount;
    this.prevFilesCount = filesCount;
    return flag;
  };

  /**
   * check cache change
   * @return {} []
   */

  _default.prototype.checkCacheChange = function checkCacheChange() {
    var autoReload = thinkCache(thinkCache.AUTO_RELOAD);
    var hasChange = false;
    for (var file in require.cache) {
      //ignore file in node_modules path
      if (file.indexOf(NODE_MODULES) > -1) {
        continue;
      }
      if (!think.isFile(file)) {
        this.clearFileCache(file);
        continue;
      }
      var mTime = _fs2['default'].statSync(file).mtime.getTime();
      if (!autoReload[file]) {
        autoReload[file] = mTime;
        continue;
      }
      if (mTime > autoReload[file]) {
        this.clearFileCache(file);
        autoReload[file] = mTime;
        hasChange = true;
        this.log(file);
      }
    }
    return hasChange;
  };

  /**
   * run
   * @return {} []
   */

  _default.prototype.run = function run() {
    var hasChange = this.checkCacheChange() || this.checkFileChange();
    if (hasChange && this.callback) {
      this.callback();
    }
    setTimeout(this.run.bind(this), 200);
  };

  /**
   * rewrite sys module load method
   * @return {} []
   */

  _default.rewriteSysModuleLoad = function rewriteSysModuleLoad() {
    var load = sys_module._load;

    //rewrite Module._load method
    sys_module._load = function (request, parent, isMain) {
      var exportsObj = load(request, parent, isMain);
      if (isMain || parent.filename.indexOf(NODE_MODULES) > -1) {
        return exportsObj;
      }
      if (request === 'internal/repl' || request === 'repl') {
        return exportsObj;
      }
      try {
        var filename = sys_module._resolveFilename(request, parent);
        var cachedModule = sys_module._cache[filename];
        if (cachedModule && parent.children.indexOf(cachedModule) === -1) {
          parent.children.push(cachedModule);
        }
      } catch (e) {}
      return exportsObj;
    };
  };

  return _default;
})(think.base);

exports['default'] = _default;
module.exports = exports['default'];
